extensions/dicom-microscopy/src/tools/microscopyManager.js

```diff
@@ -0,0 +1,591 @@
+   * @param {Object} data.roiGraphic The added ROI graphic object
+   * @param {ViewerManager} data.managedViewer The origin viewer for the event
+   */
+  _onRoiAdded(data) {
+    const { roiGraphic, managedViewer, label } = data;
+    const { studyInstanceUID, seriesInstanceUID } = managedViewer;
+    const viewState = managedViewer.getViewState();
+
+    const roiAnnotation = new RoiAnnotation(
+      roiGraphic,
+      studyInstanceUID,
+      seriesInstanceUID,
+      '',
+      viewState
+    );
+
+    this.roiUids.add(roiGraphic.uid);
+    this.annotations[roiGraphic.uid] = roiAnnotation;
+
+    roiAnnotation.subscribe(AnnotationEvents.LABEL_UPDATED, () => {
+      this.publish(EVENTS.ANNOTATION_UPDATED, roiAnnotation);
+    });
+
+    if (label !== undefined) {
+      roiAnnotation.setLabel(label);
+    } else {
+      const onRelabel = item =>
+        managedViewer.updateROIProperties({
+          uid: roiGraphic.uid,
+          properties: { label: item.label, finding: item.finding },
+        });
+      this.triggerRelabel(roiAnnotation, true, onRelabel);
+    }
+  }
+
+  /**
+   * Observes when a ROI graphic is modified, updating the correspondent
+   * annotation with the current graphic and view state.
+   *
+   * @param {Object} data The published data
+   * @param {Object} data.roiGraphic The modified ROI graphic object
+   */
+  _onRoiModified(data) {
+    const { roiGraphic, managedViewer } = data;
+    const roiAnnotation = this.getAnnotation(roiGraphic.uid);
+    if (!roiAnnotation) return;
+    roiAnnotation.setRoiGraphic(roiGraphic);
+    roiAnnotation.setViewState(managedViewer.getViewState());
+  }
+
+  /**
+   * Observes when a ROI graphic is removed, reflecting the removal in the
+   * annotations' state.
+   *
+   * @param {Object} data The published data
+   * @param {Object} data.roiGraphic The removed ROI graphic object
+   */
+  _onRoiRemoved(data) {
+    const { roiGraphic } = data;
+    this.roiUids.delete(roiGraphic.uid);
+    this.annotations[roiGraphic.uid].destroy();
+    delete this.annotations[roiGraphic.uid];
+  }
+
+  /**
+   * Observes any changes on ROI graphics and synchronize all the managed
+   * viewers to reflect those changes.
+   * Also publishes an ANNOTATION_UPDATED event to notify the subscribers.
+   *
+   * @param {Object} data The published data
+   * @param {Object} data.roiGraphic The added ROI graphic object
+   * @param {ViewerManager} data.managedViewer The origin viewer for the event
+   */
+  _onRoiUpdated(data) {
+    const { roiGraphic, managedViewer } = data;
+    this.synchronizeViewers(managedViewer);
+    this.publish(EVENTS.ANNOTATION_UPDATED, this.getAnnotation(roiGraphic.uid));
+  }
+
+  /**
+   * Creates the subscriptions for the managed viewer being added
+   *
+   * @param {ViewerManager} managedViewer The viewer being added
+   */
+  _addManagedViewerSubscriptions(managedViewer) {
+    managedViewer.subscribe(ViewerEvents.ADDED, this._onRoiAdded);
```

Comment:
in the context of OHIF, viewer is OHIF, and we have viewports. Do you meen viewport here?